pkg/storegateway/series_refs.go

```diff
@@ -379,6 +374,27 @@ func (s *mergedSeriesChunkRefsSet) Next() bool {
 	return true
 }
 
+func (s *mergedSeriesChunkRefsSet) ensureCursors(curr1, curr2 *seriesChunkRefsIterator, set1, set2 iterator[seriesChunkRefsSet]) error {
+	// When both cursors are empty, we advance their set iterators concurrently to reduce total waiting time for the
+	// IO from underlying set iterators (see grafana/mimir#4596).
+	// If either of cursors are already populated with data (or completed), the cost of goroutine switch outweigh
+	// the cost of single call to ensureItemAvailableToRead, thus below we call them sequentially.
+	if curr1.Done() && curr2.Done() {
+		var g errgroup.Group
+		g.Go(func() error { return s.ensureItemAvailableToRead(curr1, set1) })
```

Comment:
Is it safe to do concurrently? I don't remember if it was here, but I remember there was a preloading that was unsafe to do concurrently.

Reply:
I took a quick look and this is not the place I remember was unsafe to do concurrently, but please double check the actual `seriesChunkRefsIterator` implementation to make sure there's no problem here too.

Reply:
To clarify these changes: here we are pre-populating concurrently two **different** set iterators: `set1` and `set2`. I didn't find better names, but we use either `a` and `b`, or just `set`, when we refer to them within this code chunk.

Reply:
iirc these individual set iterators are just the set iterators of individual blocks. Normally all the shared resources between blocks are concurrency-safe, but may be worth double checking as Marco suggests.

Some ideas for what to check - the memory pools used to store chunks, the access to `reused` postings and series - these should all be passed in the constructors of the iterators.

Reply:
talked offline and agreed to do this with a static sleep on all Next() calls - in reality the underlying iterators would be fetching from cache, object store, and disk, so IO wait is expected on all Next() calls